---
title: "Anchorwave genome dot plots"
output: html_notebook
---

Converting raw anchorwave data into dotplots (similar to stephens excel plots) for publication.

```{r message=FALSE, warning=FALSE, include=FALSE}
library(tidyverse)
library(cowplot)
#library(ggpubr)
library(grid)
library(gridExtra)
# 
# pubTheme <-theme(title = element_text(size=10), #usually 14
#         text = element_text(size=8),
#         plot.background = element_rect(fill="lightgrey"),
#         strip.background = element_rect(linetype=0,linewidth=8,
#                                         fill="grey"), #facet boxes
#         strip.text = element_text(size=10), #usually 11
#         legend.background = element_rect(fill="grey"),
#         legend.text = element_text(size=8),
#         legend.title = element_text(size=10),
#         panel.background = element_rect(fill="#FFFFFF"), #plot body
#         panel.grid = element_blank() #plot body pt2
#   )
# pubColours <- c("#2d9da6","#99b700","#8a53b6","#bf4824","#152944")
anchors <- read_table("data/anchors",comment="#")
# select only autosomes and sex chromosomes
anchors_clean<- anchors %>% filter(queryChr == "A1" | 
                                     queryChr== "A2"|
                                     queryChr== "A4"|
                                     queryChr=="Y1"|
                                     queryChr=="Y2")
```
### plots!
```{r}
ggplot(anchors_clean,aes(referenceStart,queryStart)) + geom_point() + facet_wrap(~queryChr)
```
We want the Y to be combined and to make it easier to add PARs and stuff, we will make each plot for each chr separately.
```{r}
max_Y1<-max(filter(anchors_clean, queryChr == "Y1")$queryEnd)
max_Y2<-max(filter(anchors_clean, queryChr == "Y2")$queryEnd)

# anchors_mut <- anchors_clean %>% mutate(newstart = ifelse(queryChr == "Y2",(queryStart+max_Y1), queryStart)) %>% mutate(newChr = ifelse(queryChr == "Y1"| queryChr == "Y2","Y", queryChr)) #%>% mutate(.,queryMb = newstart/1e6) %>% mutate(.,referenceMb = referenceStart/1e6)

anchors_mut <- anchors_clean %>% #mutate(newstart = ifelse(queryChr == "Y1",(queryStart+max_Y2), queryStart)) 
  mutate(newChr = ifelse(queryChr == "Y1"| queryChr == "Y2","Y", queryChr))

anchors_mb<- anchors_mut %>% mutate(queryMb = queryStart/1e6) %>% mutate(referenceMb = referenceStart/1e6)
write_delim(anchors_mb,"data/anchors_mb.txt")

```

### separate the plots

```{r, fig.width=14, fig.height=10}
a1<- anchors_mb %>% filter(newChr=="A1") %>% 
    mutate(newstartMb=queryMb) %>%
    ggplot(aes(referenceMb,newstartMb)) + geom_point() + theme_classic() +
    theme(axis.title=element_blank())
## need to transform A2, A4 and Y so that plots are in same orientation

a2<-anchors_mb %>% filter(newChr=="A2") %>%
  mutate(newstartMb=(max(queryMb) - queryMb)) %>%
  ggplot(aes(referenceMb,newstartMb)) + geom_point() + theme_classic() +
  theme(axis.title=element_blank())

a4<-anchors_mb %>% filter(newChr=="A4") %>%
  mutate(newstartMb=(max(queryMb) - queryMb)) %>%
  ggplot(aes(referenceMb,newstartMb)) + geom_point() + theme_classic() +
  theme(axis.title=element_blank())

## modify Y plot to contain lines and colors
y<-anchors_mb %>% filter(newChr=="Y") %>%
  mutate(newstartMb=(max(queryMb) - queryMb)) %>%
  ggplot(aes(referenceMb,newstartMb)) + 
  geom_point(aes(color=queryChr)) + 
  geom_vline(xintercept = 71,linetype = 'dashed') +
  geom_vline(xintercept = 261,linetype = 'dashed') +
  geom_vline(xintercept = 367,linetype = 'dashed') +
  annotate("text", x = .35*(71), y = 400, label = "PAR1",size=5) +
  annotate("text", x = .5*(71+261), y = 400, label = "Old sex-linked\n region",size =5) +
  annotate("text", x = .5*(261+367), y = 400, label = "New\n sex-linked\n region",size=5) +
  annotate("text", x = .5*(367+500), y = 400, label = "PAR2",size=5) +
  ylim(0,430) +
  theme_classic() +
  theme(legend.position = c(0.9,0.3),axis.title=element_blank()) +
  scale_color_manual(name="",
                    labels=c("Y1","Y2"),
                     values = pubColours[c(1,4)])
  
y
z<-anchors_mb %>% filter(newChr=="Y") %>%
  mutate(newstartMb=(max(queryMb) - queryMb)) %>%
  ggplot(aes(newstartMb,referenceMb)) + 
  geom_point(aes(color=queryChr)) + 
  geom_vline(xintercept = 71,linetype = 'dashed') +
  geom_vline(xintercept = 261,linetype = 'dashed') +
  geom_vline(xintercept = 367,linetype = 'dashed') +
  annotate("text", x = .35*(71), y = 400, label = "PAR1",size=5) +
  annotate("text", x = .5*(71+261), y = 400, label = "Old sex-linked\n region",size =5) +
  annotate("text", x = .5*(261+367), y = 400, label = "New\n sex-linked\n region",size=5) +
  annotate("text", x = .5*(367+500), y = 400, label = "PAR2",size=5) +
  ylim(0,430) +
  theme_classic() +
  theme(legend.position = c(0.9,0.3),axis.title=element_blank()) +
  scale_color_manual(name="",
                    labels=c("Y1","Y2"),
                     values = pubColours[c(1,4)])
y
z
myplot<-plot_grid(a1,a2,a4,y, labels = c(" A1",
                                 " A2",
                                 " A4",
                                 "X/Y1+Y2"), 
          label_x=0.07,label_size=12,hjust=-0.3)
#create common x and y labels

y.grob <- textGrob("Maternal Haplotype\nposition(Mb)",gp=gpar(fontface="bold", fontsize=18), rot=90)

x.grob <- textGrob("Paternal Haplotype\nposition(Mb)",gp=gpar(fontface="bold", fontsize=18))

#add to plot
labeledplot<-grid.arrange(arrangeGrob(myplot, left = y.grob, bottom = x.grob))
# save as tiff (raster, lossless, good for publication apparently?)
#ggsave(plot=labeledplot,"draft4_dotplot.tiff",width=14,height=10,dpi="print",device="tiff")
```



