---
title: "Ks plots for genome paper"
output: html_notebook
---
### setup
```{r}
library(renv)
library(tidyverse)
#renv::init()
#renv::snapshot()
```

## loading in files
```{r}
ksheader<-read.table("65183_65184.CDS-CDS.last.tdd10.cs0.filtered.dag.all.go_D20_g10_A5.aligncoords.gcoords.ks.txt",comment.char ="",skip = 2,nrows = 1) %>% select(-V13) %>% rename() 
print(ksheader)
```
Tweak header names and add to main file.
```{r}
ksheader$V1<-"Ks"
ksheader$V3<-"a_dbgenomeid_chr"
ksheader$V7<-"b_dbgenomeid_chr"
as.list(ksheader)
ksfile<-read.table("65183_65184.CDS-CDS.last.tdd10.cs0.filtered.dag.all.go_D20_g10_A5.aligncoords.gcoords.ks.txt", comment.char = "#",header =FALSE)
colnames(ksfile)<- as.character(ksheader)
```
### file parsing
Change | separated entries into sep columns
```{r}
colnames(ksfile)
ksfile_split<-ksfile %>% 
      separate_wider_delim(
        cols=c("chr1||start1||stop1||name1||strand1||type1||db_feature_id1||genome_order1||percent_id1"),
          delim = "||", 
            names = c("chr1","start1_og","stop1_og","name1",
                        "strand1","type1","db_feature_id1","genome_order1","percent_id1")) %>%
      separate_wider_delim(
        cols=c("chr2||start2||stop2||name2||strand2||type2||db_feature_id2||genome_order2||percent_id2"),
          delim = "||", 
            names = c("chr2","start2_og","stop2_og","name2",
                        "strand2","type2","db_feature_id2","genome_order2","percent_id2")) 
  
head(ksfile_split)
## og start and stop are opposite of the other positions bc on minus strand
```
# tidy up 
table of just Ks values and gene IDs and chrs, filter by max 0.5 Ks. (stephen's advice) -  those genes too diverged?
```{r}
ks_df<- ksfile_split %>% 
  select(Ks,chr1,start1,start1_og,stop1,name1,chr2,start2,stop2,name2) %>%
  mutate(name1=stringr::str_replace(name1,"-RA","")) %>%
  mutate(name2=stringr::str_replace(name2,"-RA","")) %>%
  mutate(start1_mb = start1/1e6)
## annotate plot
vlines <- c(71,261,367)

ks_df_md <- ks_df %>% 
  filter(Ks< 0.5) %>%
  mutate(region = case_when((chr1 == "X" & start1_mb <=71) ~ "PAR1",
    ((chr1 == "X" & start1_mb >71 & start1_mb <=261)) ~ "OldSLR",
    ((chr1 == "X" & start1_mb >261 & start1_mb <=367)) ~ "NewSLR",
    ((chr1 == "X" & start1_mb > 367 ~ "PAR2")))) %>% 
  group_by(., region) %>% 
  mutate(mdn = median(Ks))
# standard plotting function                            
ks_plot_auto<-function(df,chr,y_lim){ggplot(filter(df,chr1 == chr), aes(start1_mb)) +
  geom_point(aes(y=Ks,alpha = 0.5)) + 
  theme_classic() +
  ylim(0,y_lim) +
  xlab(paste("position on chromosome", chr, "(Mb)"))} 
# X chr customization
ks_plot_X<- function(df,chr,y_lim){ks_plot_auto(df,chr,y_lim) +
  geom_vline(xintercept=vlines, linetype = "dashed") + 
  annotate("text", x = .35*(71), y = y_lim - 0.05, label = "PAR1",size=4) +
  annotate("text", x = .5*(71+261), y = y_lim - 0.05, label = "Old sex-linked\n region",size =4) +
  annotate("text", x = .5*(261+367), y = y_lim - 0.05, label = "New\n sex-linked\n region",size=4) +
  annotate("text", x = .5*(367+500), y = y_lim - 0.05, label = "PAR2",size=4) + 
  geom_step(aes(y = mdn, colour = "red"))}


xplot<-ks_plot_X(ks_df_md,"X",0.6)
a1<-ks_plot_auto(ks_df_md,"A1",NA)
a2<-ks_plot_auto(ks_df_md,"A2",NA)
a4<-ks_plot_auto(ks_df_md,"A4",NA)
library(cowplot)
cowplot::plot_grid(a1,a2,xplot,a4)
#ggsave("ks_genes_x.jpg",width = 6, height = 4, units = "in")
#ks_df_md_02<-ks_df_md %>% filter(Ks<0.2)
ks_plot_X((ks_df_md %>% filter(Ks<0.2)),"X",0.3)
#ggsave("ks_genes_x_02.jpg",width = 6, height = 4, units = "in")

```
### Ks in windows
```{r}
library(slider)

 windowmaker<- function(df,win){ks_df_md %>%
  group_by(chr1) %>%
  arrange(name1,chr1, start1_mb) %>%
  mutate(Ks_win = slider::slide_dbl(Ks, median, .before = win/2, .after = win/2,.step = 1,.complete =T))}
ks_win100<-windowmaker(ks_df_md,win = 100)
#ks_win50<-windowmaker(ks_df_md,win=50)

ks_plot_window<-function(df,chr,y_lim){ggplot(filter(df,chr1 == chr), aes(start1_mb)) +
  geom_line(aes(y=Ks_win)) +
  theme_classic() +
  ylim(0,y_lim) +
  xlab(paste("position on chromosome", chr, "(Mb)"))} 

y_lim<-0.08
ks_plot_window(ks_win100,"X",y_lim) + geom_vline(xintercept=vlines, linetype = "dashed") + 
  annotate("text", x = .35*(71), y = y_lim - 0.005, label = "PAR1",size=4) +
  annotate("text", x = .5*(71+261), y = y_lim - 0.005, label = "Old sex-linked\n region",size =4) +
  annotate("text", x = .5*(261+367), y = y_lim - 0.005, label = "New\n sex-linked\n region",size=4) +
  annotate("text", x = .5*(367+500), y = y_lim - 0.005, label = "PAR2",size=4) 
#ks_plot_window(ks_win50,"X",NA)
#ggsave("ks_genes_window_X_june22.jpg",dpi=300,width = 6, height = 4, units = "in")
y_lim<-0.05
a1win<-ks_plot_window(ks_win100,"A1",y_lim)
a2win<-ks_plot_window(ks_win100,"A2",y_lim)
a4win<-ks_plot_window(ks_win100,"A4",y_lim)
xwin<- ks_plot_window(ks_win100,"X",y_lim) + geom_vline(xintercept=vlines, linetype = "dashed")

cowplot::plot_grid(a1win,a2win,xwin,a4win,nrow=1)
## save as one panel

#ggsave("all_chr_ks_100win.jpg",dpi=300,width = 12, height = 4)
#ggsave(plot = a1win, "a1_ks_100win.jpg",dpi=300,width = 6, height = 4)
#ggsave(plot = a2win, "a2_ks_100win.jpg",dpi=300,width = 6, height = 4)
#ggsave(plot = a4win, "a4_ks_100win.jpg",dpi=300,width = 6, height = 4)

```
### Integrate salicifolius homology data
So I can look at Ks based on homology
```{r}
sal_raw <- data.table::fread("~/Dropbox/rumex_pangenome_annotation/pg_sal_wide_synonly.txt")

salkey<-readr::read_delim("input_to_output.txt",col_names =c("old","new"),show_col_types = FALSE)
salkey<-salkey %>% mutate(fix=str_replace(old, "\\=", "\\_")) %>%
  mutate(.,fix=str_replace(fix, "\\;", "\\_")) %>% select(fix,new)
salkey<-as.data.frame(salkey)

sal_df <- sal_raw %>% filter(salicifolius !="",hap1!="",hap2!="") %>% 
  select(interpChr,salicifolius,hap1,hap2,start) %>% mutate(start_salmb = start/1e6) %>% 
  mutate(salchr= case_when(interpChr == salkey[1,1] ~ salkey[1,2], 
                        interpChr == salkey[2,1] ~ salkey[2,2], 
                        interpChr == salkey[3,1] ~ salkey[3,2],
                        interpChr == salkey[4,1] ~ salkey[4,2],
                        interpChr == salkey[5,1] ~ salkey[5,2],
                        interpChr == salkey[6,1] ~ salkey[6,2],
                        interpChr == salkey[7,1] ~ salkey[7,2],
                        interpChr == salkey[8,1] ~ salkey[8,2],
                        interpChr == salkey[9,1] ~ salkey[9,2],
                        interpChr == salkey[10,1] ~ salkey[10,2])) %>% 
  unite(., scafpos,c("salchr","start_salmb"))


ks_df_sal <- left_join(ks_df_md, sal_df, by = c("name1"="hap1","name2"="hap2"))
windowmaker_sal<- function(df,win,start){df %>%
  group_by(salchr) %>%
  arrange(salicifolius,salchr, start) %>%
  mutate(Ks_win = slider::slide_dbl(Ks, median, .before = win/2, .after = win/2,.step = 1,.complete =T))}
orderlist<-c("Scaffold_4","Scaffold_8","Scaffold_7",
             "Scaffold_5","Scaffold_3","Scaffold_6","Scaffold_10",
             "Scaffold_1","Scaffold_2","Scaffold_9")

ks_df_sal <- transform(ks_df_sal, salchr = factor(salchr, levels = orderlist))
#levels(ks_df_sal$salchr)
ks_win100_sal<-windowmaker_sal(ks_df_sal,win = 100, start_salmb)

#ks_plot_window(ks_df_sal,"X",y_lim) + geom_rect(aes(color = interpChr))

# ggplot(filter(ks_df_sal,chr1 =="X"), aes(start1_mb)) +  
#   geom_point(aes(y=Ks_win, colour = interpChr)) +
#   theme_classic() 
# ggsave("colordot_sal_ks.jpg",dpi=300,width=6, height=4)
# ggplot(filter(ks_df_sal,chr1 =="X"), aes(start1_mb)) +  
#   geom_line(aes(y=Ks_win)) +
#   geom_rect(aes(xmin = min(start1_mb), xmax = max(start1_mb), fill = party),
#          ymin = -Inf, ymax = Inf, alpha = 0.2,
#          data = presidential) +
#   theme_classic() 
ks_plot_window_sal<-function(df,chr,y_lim){ggplot(filter(df,chr1 == chr), aes(start_salmb)) +
  geom_line(aes(y=Ks_win)) +
  theme_classic() +
  ylim(0,y_lim) +
  xlab(paste("position on chromosome", chr, "(Mb)"))} 
#head(ks_df_sal)
#ks_plot_window_sal(ks_win100_sal,"X",NA)
#ks_plot_window(ks_win100,"X",0.1) 

```
```{r}
ggplot(filter(ks_df_sal,chr1 =="X"), aes(start1_mb,group = chr1)) +  
   geom_point(aes(y=Ks, colour = salchr)) +
   theme_classic() +
  ylim(0,0.5) +
  theme(panel.spacing = unit(0,'lines'))

## order
# scaf4,scaf8,scaf7,scaf5,scaf3,scaf6,scaf10
```
```{r}
ggplot(filter(ks_df_sal,!grepl("scaffold", chr1)), aes(start1_mb)) +  
   geom_point(aes(y=Ks, colour = salchr)) +
   theme_classic() +
  ylim(0,0.5) +
  theme(panel.spacing = unit(0,'lines'))+
  facet_grid(~chr1)
```


```{r}
ggplot(filter(ks_df_sal,chr1 =="X"), aes(start_salmb,group = salchr)) +  
   geom_point(aes(y=Ks, colour = salchr)) +
   theme_classic() +
   facet_grid(~salchr) +
  ylim(0,0.5) +
  theme(panel.spacing = unit(0,'lines')) 
```
```{r}
ggplot(filter(ks_win100_sal,chr1 =="X"), aes(start_salmb,group = salchr)) +  
   geom_point(aes(y=Ks_win, colour = salchr)) +
   theme_classic() +
   facet_grid(~salchr) +
   ylim(0,0.05) +
   theme(panel.spacing = unit(0,'lines'))
```


```{r}
ggplot(filter(ks_win100_sal), aes(start_salmb,group = salchr)) +  
   geom_point(aes(y=Ks_win, colour = salchr)) +
   theme_classic() +
   facet_grid(~salchr) +
  ylim(0,0.2) +
  theme(panel.spacing = unit(0,'lines'))
```
## 
```{r}
ggplot(filter(ks_win100_sal,chr1 =="X"), aes(start_salmb,group = salchr)) +  
   geom_point(aes(y=Ks, colour = salchr,alpha=0.5)) +
  geom_line(aes(y=Ks_win)) +
   theme_classic() +
   facet_grid(~salchr) +
  ylim(0,0.2) 
ggsave("combo_Ks_salord.jpg",dpi=300,width=10,height=4)
```
Final plots for publication
```{r}
## combine into multi panel with dotplot





```





